c10/util/typeid.h

```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
 #endif
   }
 
- private:
-  const detail::TypeMetaData* data_;
+  /**
+  * convert ScalarType enum values to TypeMeta handles
+  */
+  static inline caffe2::TypeMeta fromScalarType(ScalarType scalar_type) {
+    const size_t index = static_cast<uint16_t>(scalar_type);
+    if (C10_LIKELY(index < NumScalarTypes)) {
+      return TypeMeta(index);
+    }
+    error_unrecognized_scalartype(scalar_type);
+    return TypeMeta(); // avoids "control reaches end of non-void function"
+  }
 
+  /**
+   * convert TypeMeta handles to ScalarType enum values
+   */
+  inline ScalarType toScalarType() {
+    if (C10_LIKELY(isScalarType())) {
+      return static_cast<ScalarType>(index_);
+    }
+    error_unsupported_typemeta(*this);
+    return ScalarType::Undefined; // avoids "control reaches end of non-void function"
+  }
+
+private:
+  static void error_unrecognized_scalartype(ScalarType scalar_type) {
+    TORCH_CHECK(false, "Unrecognized Scalartype ", scalar_type, " (please report this error)");
+  }
+  static void error_unsupported_typemeta(caffe2::TypeMeta dtype) {
+    TORCH_CHECK(false, "Unsupported TypeMeta in ATen: ", dtype, " (please report this error)");
+  }
```

Comment:
Wasn't your intention to hide these in a cpp file?

Reply:
It's not necessary to facilitate inlining, if that's you mean - but yeah that's where they should be.